File: May 2022/May_10.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> vec;
    
    void helper(int num, int k, int sum){
        if(num == 10){
            if(k == 0 && sum == 0){
                vector<int> v;
                v.assign(vec.begin(), vec.end());
                ans.push_back(v);
            }
            return;
        }
        
        if(num <= sum && k > 0){
            vec.push_back(num);
            helper(num + 1, k - 1, sum - num);
            vec.pop_back();
        }
        
        helper(num + 1, k, sum);
    }
    
    vector<vector<int>> combinationSum3(int k, int n) {
        helper(1, k, n);
        return ans;
    }
};

File: May 2022/May_10.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        list = new ArrayList<>();
        helper(1, k, n, new ArrayList<>());
        return list;
    }
    
    List<List<Integer>> list;
    
    private void helper(int num, int k, int sum, List<Integer> al){
        if(num == 10){
            if(sum == 0 && k == 0)
                list.add(new ArrayList<>(al));
            return;
        }
        
        if(num <= sum && k > 0){
            al.add(num);
            helper(num + 1, k - 1, sum - num, al);
            al.remove(al.size() - 1);
        }
        
        helper(num + 1, k, sum, al);
    }
}
